lib/tree.rb

```diff
@@ -1,66 +1,152 @@
-class TreeNode
-  attr_reader :key, :value
-  attr_accessor :left, :right
-
-   def initialize(key, val)
-    @key = key
-    @value = val
-    @left = nil
-    @right = nil
-   end
-end
-
-class Tree
-  attr_reader :root
-  def initialize
-    @root = nil
+  class TreeNode
+    attr_reader :key, :value
+    attr_accessor :left, :right
+  
+     def initialize(key, val)
+      @key = key
+      @value = val
+      @left = nil
+      @right = nil
+     end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  
+  class Tree 
+    attr_reader :root
+    def initialize
+      @root = nil
+    end
+  
+    # Time Complexity: 
+    # Space Complexity: 
+    def add(key, value)
+      if @root.nil?
+          @root = TreeNode.new(key,value)
+      else
+          current = @root
+          while true
+              if key < current.key
+                  if !current.left.nil?
+                      current = current.left
+                  else
+                      current.left = TreeNode.new(key,value)
+                  end
+              else
+                  if !@current.right.nil?
+                      current = current.right
+                  else
+                      current.right = TreeNode.new(key,value)
+                  end
+              end
+          end
+      end
+    end
+  
+    # Time Complexity: Because it's recursive I want to say O log(n)
+    # Space Complexity: O log (n) if it's balanced, and I guess O(n) if it isn't
+    def find(key)
```

Comment:
Because it's a binary tree the time complexity is O(log n) for a balanced tree and O(n) if it's not a balanced tree.